easier68k-cli/subcommandline_run.py

```diff
@@ -0,0 +1,225 @@
+    
+    def complete_load_memory(self, text, line, begidx, endidx):
+        return autocomplete_file(line)
+        
+    
+    
+    def do_get_memory(self, args):
+        args = split_args(args, 2, 0)
+        if(args == None):
+            return False
+            
+        memory = self.simulator.memory
+        
+        start = int(args[0], 0)
+        length = int(args[1], 0)
+        
+        # round down to nearest multiple of 8 to keep alignment
+        start -= start % 8
+        
+        for i in range(length):
+            loc = start + i
+            
+            if(loc % 8 == 0):
+                print(long_hex(loc), end='    ')
+            
+            ending = ' '
+            if(i % 4 == 3):
+                ending = '    '
+            if(i % 8 == 7):
+                ending = '\n'
+            
+            value_str = bytes(memory.get(Memory.Byte, loc)).hex()
+            print(value_str, end=ending)
+        print('') # newline
+        
+    def help_get_memory(self):
+        print('syntax: get_memory start_idx, length')
+        print('retrievs the memory in the range [start_idx, start_idx+length)')
+        print('accessing values outside of memory causes an error')
+        print('start_idx is rounded down to the nearest multiple of 8 to keep formatting alignment')
+    
+    
+    
+    def do_set_memory(self, args):
+        args = split_args(args, 3, 0)
+        if(args == None):
+            return False
+            
+        memory = self.simulator.memory
+        
+        start = int(args[0], 0)
+        length = int(args[1], 0)
+        value = args[2]
+        
+        
+        # trim off optional prefix
+        if(value[0:2] == '0x'):
+            value = value[2:]
+        
+        # check lengths
+        if(len(value) % 2 != 0 or len(value) // 2 != length):
+            print('length of value and given length do not match')
+            return False
+            
+        for i in range(length):
+            loc = start + i
+            memory.set(Memory.Byte, loc, bytearray.fromhex(value[i*2:i*2+2]))
+        
+    def help_set_memory(self, args):
+        print('syntax: get_memory start_idx, length, value')
+        print('sets the memory in the range [start_idx, start_idx+length) to the value(which must be expressed in hex)')
+        print('accessing values outside of memory causes an error')
+        print('assigning a value larger or smaller than the range can hold is an error')
+    
+    # break points when we add that too!
+    
+
+def subcommandline_run(file_name):
+    simulator = M68K()
+    if(file_name != None):
+        in_file = open(file_name)
+        
+        list_file = ListFile()
+        list_file.load_from_json(in_file.read(-1))
+        
+        in_file.close()
+        simulator.load_list_file(list_file)
+    
+    cli = Run_CLI(simulator)
+    
+    # only loops if Ctrl-C was pressed
+    while True:
+        try:
+            cli.cmdloop()
```

Comment:
will this run at 100% cpu? what do these type of applications normally do?

Reply:
Nope. it essentially looks like:
while not exit_condition:
    line = read() # here the OS will only run our OS if there is any input, otherwise it gets skipped
    do_stuff(line)